Duplicati/Library/Utility/SubStream.cs

```diff
@@ -0,0 +1,118 @@
+﻿#region Disclaimer / License
+// Copyright (C) 2015, The Duplicati Team
+// http://www.duplicati.com, info@duplicati.com
+// 
+// This library is free software; you can redistribute it and/or
+// modify it under the terms of the GNU Lesser General Public
+// License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// 
+// This library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+// Lesser General Public License for more details.
+// 
+// You should have received a copy of the GNU Lesser General Public
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+// 
+#endregion
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using System.IO;
+
+namespace Duplicati.Library.Utility
+{
+    /// <summary>
+    /// This class wraps a Stream but only allows reading a limited number of bytes from the underlying stream.
+    /// This can be used to create split a source stream into multiple smaller buffers that can be used as independent streams.
+    /// </summary>
+    public class SubStream : Stream
+    {
+        private readonly Stream m_stream;
+        private readonly long m_start;
+        private readonly long m_length;
+
+        public SubStream(Stream stream, long length)
+            : this(stream, stream.Position, length)
+        {
+        }
+
+        public SubStream(Stream stream, long start, long length)
+        {
+            this.m_stream = stream;
+            this.m_start = start;
+
+            // If the stream isn't long enough for this to be a full substream, then limit the length
+            if (stream.Length >= start + length)
+            {
+                this.m_length = length;
+            }
+            else
+            {
+                this.m_length = stream.Length - start;
+            }
+        }
+
+        public override bool CanRead => this.m_stream.CanRead;
+
+        public override bool CanSeek => this.m_stream.CanSeek;
+
+        public override bool CanWrite => false;
+
+        public override long Length => this.m_length;
+
+        public override long Position
+        {
+            get => this.m_stream.Position - this.m_start;
+            set => this.m_stream.Position = value + this.m_start;
+        }
+
+        public override void Flush()
+        {
+            this.m_stream.Flush();

```

Comment:
According to the [documentation](https://docs.microsoft.com/en-us/dotnet/api/system.io.stream.flush?view=netframework-4.7.1),

> In a class derived from Stream that doesn't support writing, Flush is typically implemented as an empty method to ensure full compatibility with other Stream types since it's valid to flush a read-only stream.

Since `CanWrite` is `false`, should we implement `Flush` to be a no-op?

Reply:
That seems reasonable to me. Updated.